### ISSUE TITLE
Visibility graphs have a huge number of polygon crossings

### ISSUE THREAD

**Issue by ekrell** [TaipanRex/pyvisgraph#48]
I am trying to create a visibility graph from a shapefile, following the examples. 
The example programs run fine, and I am able to get a correct path.
For my own shapefile, the visibility graph seem to never avoid crossing the polygons. So solution paths always cross a polygon if a polygon exists between them. 

Here is my original map, plotting the polygons with matplotlitb.
![sample_region](https://user-images.githubusercontent.com/4443660/78832564-19f6d300-79b1-11ea-9fd9-259ea6548faa.png)

For faster debugging, I made a reduced shapefile with only two polygons. 
Here, I plotted all the edges in `graph.visgraph.get_edges()`. You can also see the solution path passing through the polygon.

![sample_region_path](https://user-images.githubusercontent.com/4443660/78833219-1b74cb00-79b2-11ea-9c14-490939819bf6.png)

Any idea what might cause this? Happy to share code if it helps.
The code starts by converting a raster to polygons with fiona, saving them as a shapefile, then following example 1. Not too experienced with geospatial data, so I could have made an error here. But the polygons load fine in QGIS.

**Comment by ekrell** [comment]
Just to add some extra details, 
I remade the shapefile in QGIS in case I was doing something wrong with the python libs. I got the exact same result, so I think my shapefile is fine. 

Can the fact that this is a very high res, zoomed in region (compared to the globe-trotting example scripts) affect it? The coordinates are just not very different from point to point --> float errors?

**Comment by ekrell** [comment]
Update: coordinates seem to be at play. I scaled each by 1000 and got something better: far fewer polygon intersections!

The graph:
![sample_region_graph](https://user-images.githubusercontent.com/4443660/78843739-5a624b00-79c9-11ea-8ae4-7ec185ab202d.png)

The path:
![sample_region_path](https://user-images.githubusercontent.com/4443660/78843748-60f0c280-79c9-11ea-9239-955124516a45.png)

**Comment by ekrell** [comment]
I was able to come up with a work-around. The coastlines shapefile worked and so did my own hand-drawn test file in QGIS worked. The only one that didn't work was a raster converted to a shapefile. So I pulled up my raster->shapefile in QGIS, redrew a less "blocky" shapefile on top (still complex with lots of points).

I input this new version and it had only about 5 polygon-crossings. But after rounding and normalizing the coordinates, there were no edge crossings. 

![sample_region_graph](https://user-images.githubusercontent.com/4443660/78848892-a963ac80-79d8-11ea-9a79-1ea7a1d5653e.png)

Conclusion? The blockiness of a converted raster does not play well with the VG generation. Ultimately my pipeline will be automatically converting rasters to polygons for this, but I should be apply to smooth them a bit in the code.

**Comment by TaipanRex** [comment]
Yeah this is likely rounding issues when you have a large amount of vertices that are very close to each other. Reducing the number of vertices is the best fix, as well as scaling like you did.

**Comment by ekrell** [comment]
![sample_region_graph_mini](https://user-images.githubusercontent.com/4443660/79082027-4596f780-7ce8-11ea-8825-3398853f7c91.png)
With more data, still seeing a number of crossings. I have done numerous trials with number of rounding digits, range of new scale, etc. Note the above only shows a sample of the edges. 

More polygon smoothing might help, but they are already lower resolution than I would like. Specifically dealing with nearshore USVs. 

Anything else I should try?

**Comment by ekrell** [comment]
Just for reference, here it is with much smoother polygons. Like you said, it is sensitive to shape complexity. I would like to work with higher resolution, but I can get by with this for now. Later I can look into tuning this part. 

![sample_region_graph_mini](https://user-images.githubusercontent.com/4443660/79083086-76c7f580-7cf1-11ea-950b-7e15649d8701.png)

**Comment by SteynJanus** [comment]
Hi Ekrell. Did you already happen to find a nice solution to this? The final image seems to have no crossing but also has a high resolution. Was it just a story of finding the right scaling factor and smoothing factor? Also, did you scale it in pyvisgraph? Thanks in advance :)

**Comment by ekrell** [comment]
Hello SteynJanus,

Sure enough, my solution was just to apply a little bit of smoothing to the polygons. 

You can see the full code here where I build the shapes (some very rough code made for a class project....): 
https://github.com/ekrell/conch/blob/master/planners/vgplanner_build.py

And the exact line is here (101):   shape_ = shape_.simplify(0.0005, preserve_topology=False)

This resolution turned out fine for my needs. Hope this helps :)

**Comment by eagerbeaver04** [comment]
Hello, may be you can help me. I installed this package today and tried to get shortest path. I already tried rounding and smoothing.

![Image](https://github.com/user-attachments/assets/c69979ea-6e52-411d-be96-bac646052779)

**Comment by TaipanRex** [comment]
@eagerbeaver04 this is likely due to floating point representation issues as your coordiate system uses quite large numbers. take a look at [this documentation](https://github.com/TaipanRex/pyvisgraph/blob/documentation/docs/user_manual.rst#troubleshooting) and try reducing `COLIN_TOLERANCE` in `visible_vertices.py` (assuming you have cloned the repo), or just make your coordinates smaller by dividing by 1000.

**Comment by eagerbeaver04** [comment]
Thank you very much!

![Image](https://github.com/user-attachments/assets/40147ffc-c4d4-420e-aeec-145ade3b2349)
